### PR TITLE
fix: Duplicate function call in variable destructuring

### LINKED ISSUE #13710: [Bug]: @babel/plugin-proposal-object-rest-spread duplicates assignment calls for _objectWithoutProperties
### 💻

- [x] Would you like to work on a fix?

### How are you using Babel?

@babel/cli

### Input code

https://babeljs.io/repl#?browsers=%3E%201%25&build=&builtIns=false&corejs=false&spec=false&loose=false&code_lz=DYUwLgBA3hAeCsAuaEDu8A0EB0uCe8EAvsRALwQBmArgHYDGYAlgPa0DCAhsMACoAWnMAEkAzgBFqAB2BN6QkABMAFAEoA3EA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=true&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Ctypescript&prettier=false&targets=&version=7.13.17&externalPlugins=%40babel%2Fplugin-proposal-object-rest-spread%407.14.7&assumptions=%7B%7D

I've also made a reproducible test case in this branch https://github.com/dan-kez/babel/tree/duplicate_function_call_in_variable_destructuring

### Configuration file name

babel.config.json

### Configuration

_No response_

### Current and expected behavior

*Current Behavior:*
As shown in the repl above the `functionCallThatIsDuplicated` function is called twice in the compiled code. This is not expected as in the original code the api call is only made once.

*Expected Behavior:*
When using `@babel/plugin-proposal-object-rest-spread` right hand operation is called once and only once to match the original code.

### Environment

(Reproducible in the REPL link above)

Below is my project environment but the minimal reproducible example in the REPL link shows the same error
```
  System:
    OS: macOS 10.15.7
  Binaries:
    Node: 14.17.5 - ~/.nvm/versions/node/v14.17.5/bin/node
    Yarn: 1.22.11 - ~/.nvm/versions/node/v14.17.5/bin/yarn
    npm: 6.14.14 - ~/.nvm/versions/node/v14.17.5/bin/npm
  Monorepos:
    Yarn Workspaces: 1.22.11
    Lerna: 3.20.2
  npmPackages:
    @babel/cli: 7.12.1 => 7.12.1
    @babel/core: 7.12.3 => 7.12.3
    @babel/helper-compilation-targets: 7.12.1 => 7.12.1
    @babel/plugin-proposal-class-properties: 7.12.1 => 7.12.1
    @babel/plugin-proposal-decorators: 7.12.1 => 7.12.1
    @babel/plugin-proposal-object-rest-spread: 7.12.1 => 7.12.1
    @babel/plugin-syntax-dynamic-import: 7.8.3 => 7.8.3
    @babel/plugin-transform-react-jsx: 7.12.1 => 7.12.1
    @babel/plugin-transform-regenerator: 7.12.1 => 7.12.1
    @babel/preset-env: 7.12.1 => 7.12.1
    @babel/preset-react: 7.12.1 => 7.12.1
    @babel/preset-typescript: 7.12.1 => 7.12.1
    @babel/runtime: 7.12.1 => 7.12.1
    babel-eslint: 10.1.0 => 10.1.0
    babel-jest: 26.6.3 => 26.6.3
    babel-loader: 8.2.2 => 8.2.2
    babel-plugin-lodash: 3.3.4 => 3.3.4
    babel-plugin-module-resolver: 4.0.0 => 4.0.0
    babel-plugin-react-docgen: 4.2.1 => 4.2.1
    babel-plugin-require-context-hook: 1.0.0 => 1.0.0
    babel-plugin-styled-components: 1.11.1 => 1.11.1
    babel-plugin-transform-dynamic-import: 2.1.0 => 2.1.0
    babel-plugin-transform-react-remove-prop-types: 0.4.24 => 0.4.24
    eslint: 7.32.0 => 7.32.0
    eslint-import-resolver-babel-module: 5.3.1 => 5.3.1
    jest: 26.6.3 => 26.6.3
    lerna: 3.20.2 => 3.20.2
    webpack: 4.46.0 => 4.46.0
```

### Possible solution

I'm not familiar with the internal implementation of `@babel/plugin-proposal-object-rest-spread` but it would be possible to first assign the result of the right hand assignment to a temporary variable and then use that temporary variable in subsequent calls to `_objectWithoutProperties`

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dan-kez**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #13710 
| Patch: Bug Fix?          | Yes
| Tests Added + Pass?      | Yes
| Any Dependency Changes?  | No
| License                  | MIT

I am new to babel's code base but I believe I have a fix here for an issue where the RHS of an assignment will be called duplicate times.

I've added a few test cases to validate against this potential issue and have confirmed the output looks as I would expect.

If there are any suggestions please let me know! I'm happy to make updates.

Thanks for reviewing!

## Short summary
This PR adds a simple recursive check so that any multi property case with a spread operator on assignment will not duplicate the RHS of the assignment. 

Prior to this PR there was a guard for this case but it only looked one level deep. Example cases of this are shown in https://github.com/babel/babel/issues/13710#issuecomment-907657935

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/13711"></a>

### CODE REVIEW

File: packages/babel-plugin-proposal-object-rest-spread/test/fixtures/object-rest/variable-destructuring/input.js [hunk 1]
```diff
@@ -17,3 +17,9 @@ let {
 } = complex;
 
 let { x4: { ...y4 } } = z;
+
+let { x5: { w5, ...y5 } } = z();
+let { x6: { w6: { a6, ...y6 } } } = z();
+let { x7: { e7, r7 }, q7: { w7: { a7, ...y7 } } } = z();
+let { x8, ...y8 } = z();
+let { x9: { w9: { a9, ...y9 } }, x10: { a10, ...y10 },  } = z();
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-proposal-object-rest-spread/test/fixtures/object-rest/variable-destructuring/input.js:25]: Can you add a test case with mixed arrays?

```suggestion
let { x10: [{ w10, ...z10 }] } = z();
let { x11: [{ a11, b11 }, { c11, ...d11 }] } = z();
let { x12: [, { c12, ...d12 }] } = z();
```

Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/test/fixtures/object-rest/variable-destructuring/input.js:25]: Just pushed up a commit. Thanks for the recommendation, it caught an edge case with the implementation.

Comment (JLHwung) [reply packages/babel-plugin-proposal-object-rest-spread/test/fixtures/object-rest/variable-destructuring/input.js:25]: Can you add a new test case?
```js
const { x: [...{ x, ...y }] } = foo();
```

A RestElement's argument can be a pattern so we have to recurse into it.

Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/test/fixtures/object-rest/variable-destructuring/input.js:25]: Great point! I added the test case and confirmed the output looks as I'd expect.

Comment (JLHwung) [reply packages/babel-plugin-proposal-object-rest-spread/test/fixtures/object-rest/variable-destructuring/input.js:25]: We can not return early when `properties` is falsy, instead we should invoke the recursive function on `.argument`. For example,

```
const { x: [...{ ...y }] } = z();
```
has only one binding identifier but `doesAnyChildNodeHaveMoreThanOneProperty` returns `true`.

Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/test/fixtures/object-rest/variable-destructuring/input.js:15]: Hey @JLHwung - thanks for the the detailed case. I looked through https://astexplorer.net/ and believe I was able to diagnose the issue you called our with more robust test cases for array destructuring. 

Happy to make any other adjustments. Thank you for the thorough review and prompt responses!

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js [hunk 2]
```diff
@@ -240,6 +240,39 @@ export default declare((api, opts) => {
     }
   }
 
+  function doesAnyChildNodeHaveMoreThanOneProperty(node) {
+    if (node.argument) return true;
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/src/index.js:244]: If `node.argument` presents, it is a `RestElement` and we need to recurse to `node.argument`, otherwise we don't know if it contains more than one bindings:

```js
const [...{...a}] = []
```
When we see the first `...` after `[`, we should look into the AST and recurse to `{...a}`.

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js [hunk 3]
```diff
@@ -240,6 +240,39 @@ export default declare((api, opts) => {
...
+    // Nested object case
+    if (node.properties) {
+      if (node.properties.length > 1) return true;
+      // Recurse value of the value of the first property
+      innerNode = node.properties[0].value;
+      // Short circuit if the value is falsy
+      if (!innerNode) {
+        return false;
+      }
+    }
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/src/index.js:263]: At this point we still need to handle the assignment pattern:

```js
const { x: { ...x } = {} } = z();
```

When we recurse to `{ ...x } = {}`, we need to check the `left` (`{ ...x }`) of the assignment pattern.

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js [hunk 4]
```diff
@@ -335,7 +368,7 @@ export default declare((api, opts) => {
             // skip single-property case, e.g.
             // const { ...x } = foo();
             // since the RHS will not be duplicated
-            originalPath.node.id.properties.length > 1 &&
+            doesAnyChildNodeHaveMoreThanOneProperty(originalPath.node.id) &&
             !t.isIdentifier(originalPath.node.init)
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/src/index.js:372]: Unrelated to this PR: I think we should use `originalPath.scope.isStatic(originalPath.node.init)` here, because

1) A `ThisExpression` is safe to reuse, e.g.
```js
const { ...x } = this;
```

2) Even if `node.init` is an identifier, it can not be reused if not a constant binding. e.g.

```js
let z = { v: 1 };
function sideEffect() {
  z = { v: 2 };
}
const { x = sideEffect(), ...y } = z;
y // { v: 1 }
```
`y` becomes `{ v: 2 }` after transformed because `z` can be modified in `sideEffect()`.

These two issues will be fixed by `scope.isStatic`.

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js [hunk 5]
```diff
@@ -260,6 +265,11 @@ export default declare((api, opts) => {
       if (!innerNode) {
         return false;
       }
+
+      // If this is an AssignmentPattern, always return true else the RHS will be duplicated
+      if (innerNode.type === "AssignmentPattern" && innerNode.left) {
+        return true;
+      }
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/src/index.js:272]: Similar to `RestElement`, we need to recurse into `node.left`.

```
innerNode ??= node.left
```

Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js:272]: So I originally added that but the output did not end up as I expected. Specifically the RHS was duplicated. I believe the way that we handle assignment patterns also duplicates the RHS. 

See this diff that I branched off this branch for: https://github.com/dan-kez/babel/pull/1/files

Comment (JLHwung) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js:272]: Oh I was meant to replace the whole "if" branch with `innerNode ??= node.left`, which is similar to how we handle the array / object patterns (note that in your posted branch it was changed to `innerNode = innerNode.left`). The `innerNode` is essentially the arguments of next recursive call, so the algorithm can be summarized as

```
function hasMoreThanOneBinding(node) {
  if (node is ArrayPattern ) {
    return true if node has more than one elements
    else return hasMoreThanOneBinding(node's first element)
  } else if (node is ObjectPattern) {
    return true if node has more than one properties
    else return hasMoreThanOneBinding(node's first property)
  } else if (node is AssignmentPattern) {
    return hasMoreThanOneBinding(node's left)
  } else if (node is RestElement) {
    return hasMoreThanOneBinding(node's argument)
  }
}

Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js]: Ah I understand now. Apologies for all the back and forth here - I really appreciate you taking the time on this PR!

I've refactored the function to be more in line with you example. I did add one additional case `ObjectProperty`. This was because an ObjectPattern could have the first element a RestElement (for example `const { x15: { ...y15 } } = z();`). In this case `node.properties[0].value` would be undefined which would fall through the final else clause.

To be more specific I added an `ObjectProperty` case where I test on the presence of `node?.value` and then recurse on the given value.

If this is wrong please let me know - I'm happy to change it.


I also used optional chaining just in case I missed a case were node could be falsey (this was how I found the above issue when stepping through the test case). I am also happy to add an early return if node is undefined to return true or other cases that could be present in this LHS assignment.

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js (hunk 3 above)
Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js]: Added this test case as
```js
const { x20: { ...y20 } = { } } = z();
```

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js (hunk 2 above)
Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js]: Added the test cases below and updated as per other comments
```js
const [...[ ...y17 ]] = z();
const [...{ a19, ...y19 }] = z();
```

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js [hunk 6]
```diff
@@ -240,6 +240,31 @@ export default declare((api, opts) => {
...
+    if (node?.elements) {
+      if (node.elements.length > 1) return true;
+      else return hasMoreThanOneBinding(node.elements[0]);
+      // ObjectPattern
+    } else if (node?.properties) {
+      if (node.properties.length > 1) return true;
+      else return hasMoreThanOneBinding(node.properties[0]);
+      // ObjectProperty
+    } else if (node?.value) {
+      if (node.value.length > 1) return true;
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/src/index.js:254]:
```suggestion
```
The `value` of `ObjectProperty` is always an AST node.

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js [hunk 7]
```diff
@@ -240,6 +240,31 @@ export default declare((api, opts) => {
     }
   }
 
+  function hasMoreThanOneBinding(node) {
+    // ArrayPattern
+    if (node?.elements) {
+      if (node.elements.length > 1) return true;
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/src/index.js:246]: I just realize that `node.elements.length > 1` does not imply it has more than one binding, e.g.

```js
const [,,x] = arr
```
We should filter out the `null` nodes in `node.elements` and check that if the non-null elements have more than one element.

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js [hunk 8]
```diff
@@ -240,6 +240,31 @@ export default declare((api, opts) => {
     }
   }
 
+  function hasMoreThanOneBinding(node) {
+    // ArrayPattern
+    if (node?.elements) {
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/src/index.js:245]: After filtering out the `null` in array patterns, I actually think `node` is always non-nullish.

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js [hunk 9]
```diff
@@ -240,6 +240,31 @@ export default declare((api, opts) => {
...
+      if (node.elements.length > 1) return true;
+      else return hasMoreThanOneBinding(node.elements[0]);
+      // ObjectPattern
+    } else if (node?.properties) {
+      if (node.properties.length > 1) return true;
+      else return hasMoreThanOneBinding(node.properties[0]);
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/src/index.js:251]: we can early return `false` when `node.properties.length` is `0`, e.g.

```js
const {} = {}
```

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js (hunk 7 above)
Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js:246]: Cool I'll put up a PR for that now

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js (hunk 8 above)
Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js:245]: Sure - I'm happy to remove it but not sure if it's worth being defensive here so we don't break builds.

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js (hunk 7 above)
Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js:246]: Cool I'll put up a PR for that now

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js (hunk 9 above)
Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js:251]: I'm stepping through this and am unable to trigger this case. I believe it is because `hasMoreThanOneBinding` is called within `visitRestElements` which means we must always have at least one property somewhere in the tree if there is a rest element.

I've added the guard for safety in case the calling pattern changes outside this function but I wasn't able to trigger this case after trialing many inputs with a debugger. Similarly I wasn't able to emulate the issue in https://astexplorer.net/ if I had at least one restElement somewhere on the LHS

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js (hunk 7 above)
Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js]: Updated 😄

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js [hunk 10]
```diff
@@ -240,6 +240,34 @@ export default declare((api, opts) => {
...
+  function hasMoreThanOneBinding(node) {
+    // ArrayPattern
+    if (node.elements) {
+      const nonNullElements = node.elements.filter(element => element !== null);
+      if (nonNullElements.length > 1) return true;
+      else return hasMoreThanOneBinding(nonNullElements[0]);
+      // ObjectPattern
+    } else if (node.properties) {
+      if (node.properties.length > 1) return true;
+      // Case of `const {} = {}`. This shouldn't be hit as
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/src/index.js:252]: This branch is reachable because `visitRestElements` applies to rest elements in array patterns, too. E.g.

```js
const { x: [ ...{} ] } = { x: [] }
```

**Review by JLHwung — APPROVED** [review]
Great work!

**Comment by dan-kez** [comment]
Thanks for the review @JLHwung !! I was following along with the CI steps and noticed  [CI / E2E (create-react-app) (pull_request) was failing]https://github.com/babel/babel/runs/3495970358?check_suite_focus=true

It looks like there is in issue in how we're invoking the build command incorrectly

```
$ npm run build:prod -w react-error-overlay
npm ERR! missing script: build:prod
npm ERR! 
npm ERR! Did you mean this?
npm ERR!     build
```

I'm not sure if this error is blocking to have this fixed merged in or how I can help fixing this error.

**Comment by JLHwung** [comment]
The react e2e CI error has been fixed in #13724, you can rebase this PR on latest `main`.

**Comment by dan-kez** [comment]
> The react e2e CI error has been fixed in #13724, you can rebase this PR on latest `main`.

Perfect - thank you! Just rebased

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js [hunk 11]
```diff
@@ -240,6 +240,34 @@ export default declare((api, opts) => {
...
+      // we currently call `hasMoreThanOneBinding` only under visitRestElements.
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-proposal-object-rest-spread/src/index.js:253]:
```suggestion
```

Per https://github.com/babel/babel/pull/13711/files#r701061273

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js [hunk 12]
```diff
@@ -240,6 +240,34 @@ export default declare((api, opts) => {
     }
   }
 
+  function hasMoreThanOneBinding(node) {
+    // ArrayPattern
+    if (node.elements) {
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-proposal-object-rest-spread/src/index.js:245]: Rather than duck-checking the type of `node`, can we do something like this?

```js
if (t.isArrayPattern(node) {
} else if (t.isObjectPattern(node)) {
...
```

Comment (nicolo-ribaudo) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js:245]: "duck-checking" is inspired by "duck-typing", but probably it's not the correct term :joy:

Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js:245]: Sure I'll add that now

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js (hunk 11 above)
Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js]: Included this in the latest commit.

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js (hunk 12 above)
Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js]: Updated! I didn't realize these interface methods existed - definitely safer checks and will likely help down the line for a TS conversion.

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js [hunk 13]
```diff
@@ -240,6 +240,27 @@ export default declare((api, opts) => {
...
+      else return hasMoreThanOneBinding(node.properties[0]);
+    } else if (t.isObjectProperty(node)) {
+      return hasMoreThanOneBinding(node.value);
+    } else if (t.isAssignmentPattern(node)) {
+      return hasMoreThanOneBinding(node.left);
+    } else if (t.isRestElement(node)) {
+      return hasMoreThanOneBinding(node.argument);
+    } else {
+      // node is Identifier or MemberExpression
+      return true;
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-proposal-object-rest-spread/src/index.js:260]: Q: Since this is a single identifier, shouldn't it return false? (it's one binding)

Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js:260]: It looks like this test case fails if I make this false

input:
```
const { x15: [...{ ...y15 }] } = z();
```

output:
```
const {} = z(),
      y15 = babelHelpers.extends({}, z().x15);
```

Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js:260]: Ah I believe I have a fix. In the event  `node.argument` is an Identifier then we can early return true and this these test cases work out.

Comment (nicolo-ribaudo) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js]: Actually, I think we don't need to return `true` for rest but the fix needs to live somewhere else.

In the `const { x15: [...{ ...y15 }] } = z();` case, we don't need to cache `z()`: we can just skip the `{} = z()` one.

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js [hunk 14]
```diff
@@ -240,6 +240,29 @@ export default declare((api, opts) => {
     }
   }
 
+  function hasMoreThanOneBinding(node) {
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/src/index.js:243]: Since this function is independent and complex, we could extract it to a new module and create unit tests for it.

The input can be created from source => (babel parser) => ast, the test function is `hasMoreThanOneBinding`. It will be easier to reason about whether `hasMoreThanOneBinding` works as expected than going through the messy fixture outputs.

See https://github.com/babel/babel/blob/main/packages/babel-types/test/retrievers.js for an example test setup.

Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js:243]: I have something like this working.

would you be amenable to something like this?

```js
import { parse } from "@babel/parser";
import hasMoreThanOneBinding from "../lib/hasMoreThanOneBinding";

function getBody(program) {
  return parse(program, { sourceType: "module" }).program.body;
}
describe("hasMoreThanOneBinding", function () {
  it.each([
    [
      "variable declarations",
      getBody("const { x15: [...{ ...y15 }] } = z();"),
      false,
    ],
  ])("%s", (_, ast, expectedResult) => {
    const result = hasMoreThanOneBinding(ast);
    expect(result).toEqual(expectedResult);
  });
});

```

Happy to do this. I would like to get this fix in sooner as we've had a few regressions due to this bug at work.

Comment (JLHwung) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js:243]: That looks great! You may also put `getBody` in the test functions so the test cases can be
```
[ "const { x15: [...{ ...y15 }] } = z();", false ],
[ "const { x15: [...{ ...y15, z15 }] } = z();", true ],
...
```

Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js]: @JLHwung made a distinct test file with these test cases. I copied the test cases from input.js and added the expected boolean return.

File: packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+import hasMoreThanOneBinding from "../lib/hasMoreThanOneBinding";
+
+function getFistObjectPattern(program) {
+  return parse(program, { sourceType: "module" }).program.body[0]
+    .declarations[0].id;
+}
+describe("hasMoreThanOneBinding", function () {
+  it.each([
+    ["const { x: { ...y } } = z();", true],
+    ["let { x4: { ...y4 } } = z();", true],
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js:11]: It should return `false`, the source contains only one binding `y4`. Also we don't have to add number suffix since every snippet is parsed independently.

File: packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+    ["let { x7: { e7, r7 }, q7: { w7: { a7, ...y7 } } } = z();", true],
+    ["let { x8, ...y8 } = z();", true],
+    ["let { x9: { w9: { a9, ...y9 } }, x10: { a10, ...y10 },  } = z();", true],
+    ["let { x11: [{ w11, ...z11 }] } = z();", true],
+    ["let { x12: [{ a12, b12 }, { c12, ...d12 }] } = z();", true],
+    ["let { x13: [, { c13, ...d13 }] } = z();", true],
+    ["const { x14: [...{ q14, ...y14 }] } = z();", true],
+    ["const { x15: [...{ ...y16 }] } = z();", true],
+    ["const [...[ ...y17 ]] = z();", true],
+    ["const [...{ ...y18 }] = z();", true],
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js:23]: These two cases are almost identical, we can remove one of them.

File: packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js [hunk 17]
```diff
@@ -0,0 +1,46 @@
+import { parse } from "@babel/parser";
+import hasMoreThanOneBinding from "../lib/hasMoreThanOneBinding";
+
+function getFistObjectPattern(program) {
+  return parse(program, { sourceType: "module" }).program.body[0]
+    .declarations[0].id;
+}
+describe("hasMoreThanOneBinding", function () {
+  it.each([
+    ["const { x: { ...y } } = z();", true],
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js:10]: Should return `false`.

File: packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+    ["let { x5: { w5, ...y5 } } = z();", true],
+    ["let { x6: { w6: { a6, ...y6 } } } = z();", true],
+    ["let { x7: { e7, r7 }, q7: { w7: { a7, ...y7 } } } = z();", true],
+    ["let { x8, ...y8 } = z();", true],
+    ["let { x9: { w9: { a9, ...y9 } }, x10: { a10, ...y10 },  } = z();", true],
+    ["let { x11: [{ w11, ...z11 }] } = z();", true],
+    ["let { x12: [{ a12, b12 }, { c12, ...d12 }] } = z();", true],
+    ["let { x13: [, { c13, ...d13 }] } = z();", true],
+    ["const { x14: [...{ q14, ...y14 }] } = z();", true],
+    ["const { x15: [...{ ...y16 }] } = z();", true],
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js:21]: Should return `false`.

File: packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+    ["const [...[ ...y17 ]] = z();", true],
+    ["const [...{ ...y18 }] = z();", true],
+    ["const [...{ a19, ...y19 }] = z();", true],
+    ["const { x20: { ...y20 } = { } } = z();", true],
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js:25]: Should return `false`.

File: packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+    ["const { x22: { q22, ...y22  } = {} } = z();", true],
+    ["const [[ ...y23 ] = []] = z();", true],
+    ["const [{ ...y24 } = []] = z();", true],
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js:28]: Should return `false` and almost identical examples.

File: packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+    ["const { x25: [ ...y25 ] = []} = z();", true],
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js:29]: Should return `false`.

File: packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+    ["const { x26: [ q26, ...y26 ] = []} = z();", true],
+    ["const {x28: [,,{...y28}]} = z();", true],
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js:31]: Should return `false`

File: packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+    ["const {x29: [,,{q29, ...y29}]} = z();", true],
+    ["const [,,{y30, ...x30}] = z();", true],
+    ["const [,,{...x31}] = z();", true],
+    ["const { x32: { }, w32: { ...y32 } } = z();", true],
+    ["const [,,{}, {...q32}] = z();", true],
+    ["const { ...y33 } = z();", true],
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js:37]: These four examples also contains only one binding, and thus should return `false`.

File: packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+    ["const { x22: { q22, ...y22  } = {} } = z();", true],
+    ["const [[ ...y23 ] = []] = z();", true],
+    ["const [{ ...y24 } = []] = z();", true],
+    ["const { x25: [ ...y25 ] = []} = z();", true],
+    ["const { x26: [ q26, ...y26 ] = []} = z();", true],
+    ["const {x28: [,,{...y28}]} = z();", true],
+    ["const {x29: [,,{q29, ...y29}]} = z();", true],
+    ["const [,,{y30, ...x30}] = z();", true],
+    ["const [,,{...x31}] = z();", true],
+    ["const { x32: { }, w32: { ...y32 } } = z();", true],
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js:35]: This is a very interesting case. It shows that we can't simply return `true` if `node.properties > 1`, since the property value may be an empty object pattern, _or_ nested pattern.

We may need a new algorithm similar to

https://github.com/babel/babel/blob/5fcf876c5761f12af47facdfc994b73e842cf1df/packages/babel-types/src/retrievers/getBindingIdentifiers.ts#L34

Essentially the `hasMoreThanOneBinding` should be equivalent to
```
Object.keys(getBindingIdentifiers(node)).length > 1
```

File: packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js (hunk 15 above)
Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/test/hasMoreThanOneBinding.test.js:11]: So I think `hasMoreThanOneBinding` is an invalid name fo the function. The end goal of this is to strictly stop the right hand side being duplicated

I am happy to rename this function or do something else but in the end the main goal is to mitigate the active bug.

File: packages/babel-plugin-proposal-object-rest-spread/src/index.js (hunk 13 above)
Comment (dan-kez) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js]: Would it be ok to consider making this change in a separate PR as an optimization? My main goal for this is to mitigate duplicating the right hand assignment. I'd prefer avoiding doing a larger refactor of the greater plugin if possible.

Comment (nicolo-ribaudo) [reply packages/babel-plugin-proposal-object-rest-spread/src/index.js]: Yes, that's ok!

**Comment by dan-kez** [comment]
Hey @JLHwung & @nicolo-ribaudo 

I'd love to align on the implementation that you both would like to see to have this bug fixed.

Right now I would say the acceptance criteria should likely be the following:
1. There are no instances of the RHS of an assignment being duplicated due to variable destructuring.
1. The code is readable and tested.
1. Optimizations where we intelligently determine if we would need a temporary variable can be done in a separate PR with a distinct issue to track it. This is to minimize scope of this ticket.

To accomplish the above I want to do the following
1. Create a new issue to track the optimization listed in #3.
1. Rename `hasMoreThanOneBinding` to `shouldStoreRHSInTemporaryVariable` with a comment linking to this discussion and the issue made above.

Would making both those changes be sufficient to see this PR merged?

**Comment by nicolo-ribaudo** [comment]
Since this PR fixes the bug in the generated code, I'm ok with just renaming `hasMoreThanOneBinding` and merging it.

**Comment by nicolo-ribaudo** [comment]
Could you rebase and run `yarn && yarn dedupe` to fix the CI failure?

**Comment by dan-kez** [comment]
Thanks for the prompt response @nicolo-ribaudo !

I updated the function name, added a comment, and also ran yarn dedupe.

**Review by nicolo-ribaudo — APPROVED** [review]
Let's also wait for @JLHwung's opinion.

**Review by JLHwung — APPROVED** [review]
This PR already fixed several cases. We can iterate later on further issues.

**Comment by dan-kez** [comment]
Excellent! Thank you both!I took a look at CI and I'm not sure if there is something a maintainer / member needs to do to enable the tests to continue.

**Comment by JLHwung** [comment]
@dan-kez All good! Merged.

**Comment by dan-kez** [comment]
Awesome! Thank you @JLHwung - looking forward to incorporating this at to work once a new version is cut. 

Really appreciate the thorough review from both of you!
